app/lib/storage/encryption.ts

```diff
@@ -0,0 +1,159 @@
+import * as crypto from "crypto"
+
+/**
+ * This module contains all the functions related to encryption middleware for
+ * storage.
+ */
+
+/**
+ * This interface represents the arguments of the Options class constructor.
+ */
+export interface IProtoOptions {
+  algorithm?: string
+  digest?: string
+  iterations?: number
+  iv?: Buffer
+  ivBytes?: number
+  keyBytes?: number
+  password: string
+  salt?: Buffer
+  saltBytes?: number
+}
+
+/**
+ * The Options class bundles together all cryptographic settings and methods for
+ * a vault.
+ */
+export class Options {
+  /** The name of a cryptographic algorithm */
+  public algorithm: string
+  /** The name of a hash function */
+  public digest: string
+  /** Number of PBKDF2 iterations */
+  public iterations: number
+  /** PBKDF2 initialization vector */
+  public iv: Buffer
+  /** PBFDK2 initialization vector byte length */
+  public ivBytes: number
+  /** AES key byte length */
+  public keyBytes: number
+  /** User defined password */
+  public password: string
+  /** PBFDK2 salt */
+  public salt: Buffer
+  /** PBFDK2 salt byte length */
+  public saltBytes: number
+
+  constructor(opts: IProtoOptions) {
+    this.algorithm = opts.algorithm || "aes-256-cbc"
+    this.digest = opts.digest || "sha256"
+    this.iterations = opts.iterations || 100000
+    this.ivBytes = opts.ivBytes || 16
+    this.keyBytes = opts.keyBytes || 32
+    this.password = opts.password
+    this.saltBytes = opts.saltBytes || 32
+
+    this.iv = opts.iv || crypto.randomBytes(this.ivBytes)
+    this.salt = opts.salt || crypto.randomBytes(this.saltBytes)
```

Comment:
By that point (line 56), `this.ivBytes` and `this.saltBytes` just can't be `null` as they are ORed against default values earlier in the constructor. This code is actually safer than it may seem at first sight. 

However, I understand that your concern is about the constructor being modified in the future in any way that could break the type safety of `this.ivBytes` and `this.saltBytes`, for example if they lose their default values.